src/dag.rs

```diff
@@ -30,43 +174,414 @@ impl<Types: RepoTypes> IpldDag<Types> {
+        };
+
+        let mut iter = path.iter().peekable();
+
+        let (node, _) = match self.resolve0(cid, &mut iter, true).await {
+            Ok(t) => t,
+            Err(e) => {
+                drop(iter);
+                return Err(e.with_path(path));
+            }
         };
-        let mut ipld = decode_ipld(&cid, self.ipfs.repo.get_block(&cid).await?.data())?;
-        for sub_path in path.iter() {
-            ipld = try_resolve(ipld, sub_path)?;
-            ipld = match ipld {
-                Ipld::Link(cid) => decode_ipld(&cid, self.ipfs.repo.get_block(&cid).await?.data())?,
-                ipld => ipld,
+
+        Ipld::try_from(node)
+    }
+
+    /// Resolves a Cid rooted path to a document "node."
+    ///
+    /// The return value has two kinds of meanings depending on whether links should be followed or
+    /// not: When following links the second returned value will be the path inside last document.
+    /// When not following links the second returned value will be the unmatched or "remaining"
+    /// path.
+    ///
+    /// Regardless of the `follow_links` option, HAMT sharded directories will be resolved through
+    /// as a "single step" in the given IpfsPath.
+    ///
+    /// Returns the node and remaining path or the path inside the last document.
+    pub async fn resolve(
+        &self,
+        path: IpfsPath,
+        follow_links: bool,
+    ) -> Result<(ResolvedNode, SlashedPath), ResolveError> {
+        // FIXME: do ipns resolve first
+        let cid = match path.root().cid() {
+            Some(cid) => cid,
+            None => panic!("not implemented: Ipns resolution, expected Cid based path"),
+        };
+
+        let (node, matched_segments) = {
+            let mut iter = path.iter().peekable();
+            match self.resolve0(cid, &mut iter, follow_links).await {
+                Ok(t) => t,
+                Err(e) => {
+                    drop(iter);
+                    return Err(e.with_path(path));
+                }
+            }
+        };
+
+        // we only care about returning this remaining_path with segments up until the last
+        // document but it can and should contain all of the following segments (if any). there
+        // could be more segments when `!follow_links`.
+        let remaining_path = path.into_shifted(matched_segments);
+
+        Ok((node, remaining_path))
+    }
+
+    /// Return the node where resolving ended, and the **count** of segments matched.
+    async fn resolve0<'a>(
+        &self,
+        cid: &Cid,
+        segments: &mut Peekable<impl Iterator<Item = &'a str>>,
+        follow_links: bool,
+    ) -> Result<(ResolvedNode, usize), RawResolveLocalError> {
+        use LocallyResolved::*;
+
+        let mut current = cid.to_owned();
+        let mut total = 0;
+
+        let mut cache = None;
+
+        loop {
+            let block = match self.ipfs.repo.get_block(&current).await {
```

Comment:
the fact that the resolving is "local" immediately made me think "why isn't this `get_block_now`, but then I remembered the earlier comment on how this name is misleading, which I definitely agree with :smile:.

Reply:
Yeah it was quite confusing but implementation wise it does make sense, it allows to do non-async test cases which is always a win, even if they are now quite verbose.